extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -258,6 +413,115 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
+            // set sra request promise which results in sending as
+            // non-SRA request (benefit is it allows direct cache method).
+            if (typeInstances.length == 1) {
+              dev().info(TAG, `single block in network ${networkId}`);
+              typeInstances[0].sraResponseResolver_(null);
+              return;
+            }
+            // Construct and send SRA request.
+            // Chunk hanlder called with metadata and creative for each slot
+            // in order of URLs given.  Construct promise for each slot
+            // such that its resolver will be called.
+            const sraRequestAdUrlResolvers =
+              typeInstances.map(instance => instance.sraResponseResolver_);
+            // TODO(keithwrightbos) - how do we handle per slot 204 response?
+            return constructSRARequest_(
+                this.win, this.getAmpDoc(), typeInstances)
+              .then(sraUrl => {
+                checkStillCurrent();
+                return xhrFor(this.win).fetch(sraUrl, {
+                  mode: 'cors',
+                  method: 'GET',
+                  credentials: 'include',
+                });
+              })
+              .then(response => lineDelimitedStreamer(this.win, response,
+                metaJsonCreativeGrouper((creative, headers) => {
+                  checkStillCurrent();
+                  // Force safeframe rendering method.
+                  headers[RENDERING_TYPE_HEADER] = XORIGIN_MODE.SAFEFRAME;
+                  // Need to convert to FetchResponse object?
+                  const fetchResponseHeaders =
+                    /** @type {?../../../src/service/xhr-impl.FetchResponseHeaders} */
+                    ({
+                      get: name => headers[name],
+                      has: name => !!headers[name],
+                    });
+                  const fetchResponse =
+                    /** @type {?../../../src/service/xhr-impl.FetchResponse} */
+                    ({
+                      headers: fetchResponseHeaders,
+                      arrayBuffer: () => utf8Encode(creative),
+                    });
+                  sraRequestAdUrlResolvers.shift()(fetchResponse);
+                })))
+              .catch(error => {
+                const canceled = isCancellation(error);
+                if (!canceled) {
+                  user().error(TAG, 'SRA request failure', error);
+                }
+                // Collapse all slots on failure so long as they are not
+                // cancellation.
+                typeInstances.forEach(instance => {
+                  // Reset ad url to ensure layoutCallback does not fallback to
+                  // frame get which would lose SRA guarantees.
+                  // TODO(keithwrightbos): publisher should indicate if
+                  // explicit is required!
+                  instance.resetAdUrl();
+                  if (!canceled) {
+                    instance.forceCollapse();
```

Comment:
`#forceCollapse` should be a privileged API (causes page jumps). Isn't there one that collapses conditionally like `#attemptChangeHeight`?